In [1]:
import pandas as pd
import numpy as np
import os

### Importing Data Sets

In [2]:
path = r'C:\Users\facun\Desktop\Data Analysis\CF\PYTHON\Instacart Basket Analysis'

In [4]:
order_prods_prior = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders_products_prior.csv'),index_col = False)

In [10]:
customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'),index_col = False)

## Consistency check of "order_prods_prior" file

In [6]:
order_prods_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [7]:
order_prods_prior.shape

(32434489, 4)

### Checking for missing values

In [8]:
order_prods_prior.isnull().sum()

order_id             0
product_id           0
add_to_cart_order    0
reordered            0
dtype: int64

#### Answer: no missing values found

### Checking for Duplicates

In [9]:
df_dups = order_prods_prior[order_prods_prior.duplicated()]
df_dups

Empty DataFrame
Columns: [order_id, product_id, add_to_cart_order, reordered]
Index: []

#### Answer: no duplicates found

## Consistency check of "customers" file 

In [11]:
customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

### Checking for missing values

In [12]:
customers.isnull().sum()

user_id             0
First Name      11259
Surnam              0
Gender              0
STATE               0
Age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

#### Answer: 11259 missing values of the column First Name. No changes made as it doesn't affect the analysis of the date

### Checking for Duplicates

In [13]:
df_dups_cus = customers[customers.duplicated()]
df_dups_cus

Empty DataFrame
Columns: [user_id, First Name, Surnam, Gender, STATE, Age, date_joined, n_dependants, fam_status, income]
Index: []

#### Answer: No duplicates found

### Merging "customers" file and "orders_products_merged"

In [14]:
# Importing the last orders_products_merged

In [18]:
orders_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_prods_merge.pkl'))

In [19]:
customers.columns

Index(['user_id', 'First Name', 'Surnam', 'Gender', 'STATE', 'Age',
       'date_joined', 'n_dependants', 'fam_status', 'income'],
      dtype='object')

In [20]:
customers.shape

(206209, 10)

In [21]:
orders_prods_merge.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_last_order', 'product_id',
       'add_to_cart_order', 'reordered', 'product_name', 'aisle_id',
       'department_id', 'prices', '_merge', 'price_range_loc', 'businest_day',
       'businest_period_of_day', 'max_order', 'loyalty_flag', 'average_price',
       'spending_flag', 'median_days_since_last_order',
       'order_frequency_flag'],
      dtype='object')

In [27]:
orders_prods_merge = orders_prods_merge.drop(columns=['_merge'])

In [28]:
orders_prods_merge.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_last_order', 'product_id',
       'add_to_cart_order', 'reordered', 'product_name', 'aisle_id',
       'department_id', 'prices', 'price_range_loc', 'businest_day',
       'businest_period_of_day', 'max_order', 'loyalty_flag', 'average_price',
       'spending_flag', 'median_days_since_last_order',
       'order_frequency_flag'],
      dtype='object')

In [29]:
orders_prods_merge.shape

(32406041, 22)

In [39]:
orders_products_all = orders_prods_merge.merge(customers, on = 'user_id', indicator = True)

In [40]:
orders_products_all.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  ...  First Name  Surnam  Gender    STATE Age date_joined  \
0         Soda  ...       Linda  Nguyen  Female  Alabama  31   2/17/2019   
1         Soda  ...       Linda  Nguyen  Female  Alabama  31   2/17/2019   
2         Soda  ...       Linda  Nguyen  Female  Alabama  31   2/17/2019   
3         Soda  ...       Linda  Nguyen  Female  Alabama  31   2/17/2019   
4         Soda  ...       Linda  Nguyen  Female  Alabama  31   2/17/2019   

   n_dependants fam_status  income _merge  
0             3    married   40423   both  
1             3    married   40423   both  
2             3    married   40423   both  
3             3    married   40423   both  
4             3    married   40423   both  

[5 rows x 32 columns]

In [41]:
orders_products_all.shape

(32406041, 32)